planner/core/exhaust_physical_plans.go

```diff
@@ -1271,10 +1271,30 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *property.PhysicalProperty) (indexJ
 
 	forceLeftOuter := inljLeftOuter || inlhjLeftOuter || inlmjLeftOuter
 	forceRightOuter := inljRightOuter || inlhjRightOuter || inlmjRightOuter
+	needForced := forceLeftOuter || forceRightOuter
 
 	defer func() {
 		// refine error message
-		if !forced && (hasINLJHint || hasINLHJHint || hasINLMJHint) {
+		if !canForced && needForced {
+			if hasINLMJHint && len(indexJoins) > 0 && len(prop.Items) > 0 {
+				containIdxMergeJoin := false
+				for _, idxJoin := range indexJoins {
+					if _, ok := idxJoin.(*PhysicalIndexMergeJoin); ok {
+						containIdxMergeJoin = true
+						break
+					}
+				}
+				// Index merge join has more strict conditions than index join and index hash join,
```

Comment:
1. IndexMergeJoin requires stricter conditions than Index(Hash)Join when the output order is needed.
2. IndexMergeJoin requires looser conditions than Index(Hash)Join when the output is unordered.
3. If ordered-Index(Hash)Join can be chosen but ordered-IndexMergeJoin can not be chosen, we can build a plan with an enforced sort on IndexMergeJoin.
4. Thus we can give up the plans here if IndexMergeJoin is nil when `hasINLMJHint` is true. Because we can make sure that an IndexMeregJoin with enforced sort will be built.